File: src/main/java/com/udalmik/algs/interviewbit/backtracking/GenerateParenthesis.java
package com.udalmik.algs.interviewbit.backtracking;

import java.util.ArrayList;
import java.util.Collections;

public class GenerateParenthesis {

    public ArrayList<String> generateParenthesis(int n) {
        ArrayList<String> result = new ArrayList<>();
        generateParenthesisInternal(n, 0, 0, "", result);
        Collections.sort(result);
        return result;
    }

    private void generateParenthesisInternal(int n, int left, int right, String current, ArrayList<String> result) {
        if (current.length() > n * 2) {
            return;
        }
        if (current.length() == n * 2 && left == right) {
            result.add(current);
        }
        if (left < n) {
            generateParenthesisInternal(n, left + 1, right, current + "(", result);
        }
        if (right < left) {
            generateParenthesisInternal(n, left, right + 1, current + ")", result);
        }
    }

}

File: js/leetcode/backtracking/generate-parenthesis.js


const generateParenthesis = n => {
    return _generateParenthesis([""], n, n);
};

const _generateParenthesis = (current, left, right) => {
    if (left <= 0 && right <= 0) {
        return current;
    }
    let result = [];
    if (left > 0) {
        // append (
        result = _generateParenthesis(current.map(curr => curr + "("), left - 1, right);
    } 
    if (right > left) {
        result = [
            ...result, 
            ..._generateParenthesis(current.map(c => c + ")"), left, right - 1)
        ]
    }
    return result;
}

module.exports = generateParenthesis;